crates/bevy_ui/src/layout_components.rs

```diff
@@ -0,0 +1,430 @@
+//! Components used to control the layout of UI-[`Node`](super::Node) entities.
+use crate::{Size, UiRect, Val};
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::prelude::Component;
+use bevy_reflect::prelude::*;
+use serde::{Deserialize, Serialize};
+
+use self::flex::FlexLayout;
+
+/// Controls which layout algorithm is used to position this UI node
+#[derive(Component, Copy, Clone, PartialEq, Debug, Serialize, Deserialize, Reflect)]
+#[reflect_value(PartialEq, Serialize, Deserialize)]
+pub enum Layout {
+    /// Use the absolute position and size specified.
+    PositionLayout,
+    /// Use the [Flexbox](https://cssreference.io/flexbox/) layout algorithm.
+    FlexLayout(FlexLayout),
+}
+
+impl Default for Layout {
+    fn default() -> Self {
+        Layout::FlexLayout(FlexLayout::default())
+    }
+}
+
+/// The strategy used to position this node.
+#[derive(
+    Component, Copy, Clone, PartialEq, Eq, Debug, Default, Serialize, Deserialize, Reflect,
+)]
+#[reflect_value(PartialEq, Serialize, Deserialize)]
+pub enum PositionType {
+    /// Relative to all other nodes with the [`PositionType::Relative`] value.
+    #[default]
+    Relative,
+    /// Positioned in relation to its parent,
+    /// but calculated as if this was the only child of its parent.
+    Absolute,
+}
+
+/// The offset of a UI node from its base position
+///
+/// Layout is performed according to the [`Layout`], and then this value is added at the end.
+/// When this is [`Layout::PositionLayout`], this value will represent the absolute position of the UI node.
+/// To check the final position of a UI element, read its [`Transform`](bevy_transform::components::Transform) component.
+#[derive(
+    Component,
+    Deref,
+    DerefMut,
+    Copy,
+    Clone,
+    PartialEq,
+    Debug,
+    Default,
+    Serialize,
+    Deserialize,
+    Reflect,
+)]
+#[reflect_value(PartialEq, Serialize, Deserialize)]
+pub struct Offset(pub UiRect);
+
+/// Controls the size of UI nodes
+///
+/// Layout is performed according to the [`Layout`]
+/// To check the actual size of a UI element, read its [`Transform`](bevy_transform::components::Transform) component
+#[derive(Component, Copy, Clone, PartialEq, Debug, Default, Serialize, Deserialize, Reflect)]
+#[reflect_value(PartialEq, Serialize, Deserialize)]
+pub struct SizeConstraints {
+    /// The minimum extent, which cannot be violated by the layouting algorithm
+    pub min: Size,
+    /// The suggested extent, which will be used if other constraints can be comfortably satisfied
+    pub suggested: Size,
+    /// The maximum extent, which cannot be violated by the layouting algorithm
+    pub max: Size,
+    /// The expected aspect ratio, computed as width / height
```

Comment:
It sounds like the layout is allowed to violate this? Is that really the case?

Reply:
I'm not sure what the exact nature of Taffy's aspect-ratio is atm. I know we want to support various methods for calculating aspect-ratio: width-defines-height, height-defines-width ++, so I suggest leaving this for now when updating this and it's message once we expand on the aspect ratio later.